### ISSUE TITLE
Media file link in text field conversion to asset

### ISSUE THREAD

**Issue by tkrch** [Kentico/xperience-by-kentico-kentico-migration-tool#128]
documentation update related to #125

---
## TL;DR
add this to configuration to activate feature for conversion of field "text" with media link to field of type "asset":
```json
"OptInFeatures":{
  "CustomMigration":{
    "FieldMigrations": [
      {
        "SourceDataType": "text",
        "TargetDataType": "assets",
        "SourceFormControl": "MediaSelectionControl",
        "TargetFormComponent": "Kentico.Administration.AssetSelector",
        "Actions": [ "convert to asset" ],
        "FieldNameRegex": ".*"
      }
    ]
  }
}   
```

---

XbK currently not supports media selection for text fields. to mitigate this issue solution was devised.

## Motivation

Instead migrating links as strings to target XbK instance, conversion to asset datatype is made. This is useful when support for media updates on UI is required through "Media selection" component (otherwise only textbox is supported).

## Supported link formats

- ✔️ [Permanent](https://docs.xperience.io/configuring-xperience/configuring-the-environment-for-content-editors/configuring-media-libraries)
  - ✔️permanent media file `~/getmedia/CCEAD0F0-E2BF-459B-814A-36699E5C773E/somefile.jpeg?width=300&height=100`
  - ✔️permanent attachment `~/getattachment/CCEAD0F0-E2BF-459B-814A-36699E5C773E/somefile.jpeg?width=300&height=100`
  - ✔️ attachment with node alias path `"~/getattachment/NicePath/To/mypage/somefile.jpeg?width=300&height=100`
- ❌[Direct file path](https://docs.xperience.io/configuring-xperience/configuring-the-environment-for-content-editors/configuring-media-libraries)
  - ❌ `~/mysitename/media/mylibraryname/sub-dir/sub-sub-dir/somefile.png?ext=.png`

## Pitfall of this approach

this approach will require changes to some consumers of field data (where data is consume literally). Field will no longer contain link in format `~/getmedia/CCEAD0F0-E2BF-459B-814A-36699E5C773E/somefile.jpeg?width=300&height=100`, but asset datatype compatible structure, for example `[{"Identifier":"e61fb752-03d5-4a03-ac0f-f187e690611e","Name":"mymediafile","Size":11803,"Dimensions":{"Width":912,"Height":202}}]`

## Opt-in feature configuration

Feature is configured in following way using `appsettings.json`:
```json
"OptInFeatures":{
  "CustomMigration":{
    "FieldMigrations": [
      {
        "SourceDataType": "text",
        "TargetDataType": "assets",
        "SourceFormControl": "MediaSelectionControl",
        "TargetFormComponent": "Kentico.Administration.AssetSelector",
        "Actions": [ "convert to asset" ],
        "FieldNameRegex": ".*"
      }
    ]
  }
}   
```

`CustomMigration` - container for custom / non-standard way of migration, declarative customization of process
`FieldMigrations` - array of directives for specification of field and datatype migration

`FieldMigration` object:
Default implementation can be seen in file `Migration.Toolkit.Core/Services/CmsClass/FormFieldMappingModel.cs` property `BuiltInFieldMigrations`.

- `SourceDataType` KX13 field datatype
- `TargetDataType` XbK target field datatype
- `SourceFormControl` KX13 form control name
- `TargetFormComponent` XbK form component
- `Actions` additional actions to be applied, possible values for now
  - `convert to asset`
  - `convert to pages`
- `FieldNameRegex` field migration rule will be applied only when field name matches this regex `.*` for all fields. Regex is case insensitive

### Without custom conversion
![image](https://user-images.githubusercontent.com/23259342/223593737-3966061b-c70d-4a48-8383-813df427a3d9.png)

### With custom conversion
![image](https://user-images.githubusercontent.com/23259342/223594012-73f90a2d-3042-4102-857f-776aad1137a2.png)

**Comment by DavidBKentico** [comment]
Added to the Kentico TW backlog. We'll take a look as soon as we can.

**Comment by skurekjakub** [comment]
Closing.
